### PR TITLE
add constant_exprt variants of numeric_cast

### PR DESCRIPTION
**Pull request by kroening**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/arith_tools.h [hunk 1]
```diff
@@ -45,11 +45,20 @@ template <>
...
+      return {};
+    else
+      return operator()(to_constant_expr(expr));
+  }
+
+  optionalt<mp_integer> operator()(const constant_exprt &expr) const
   {
     mp_integer out;
-    if(expr.id() != ID_constant || to_integer(to_constant_expr(expr), out))
+    if(to_integer(to_constant_expr(expr), out))
```
Comment (tautschnig) [inline src/util/arith_tools.h]: `to_constant_expr` isn't necessary anymore.

**Review by tautschnig — COMMENTED** [review]
What is the desired benefit? (Commit message is lacking this info...)

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c8be185a5e8196880ebdbf0abfd8e62517b88ba0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99177767

**Review by smowton — APPROVED** [review]
Except as per @tautschnig

File: src/util/arith_tools.h (hunk 1 above)
Comment (kroening) [reply src/util/arith_tools.h]: fixed

**Review by tautschnig — APPROVED** [review]
I find "numeric_cast is frequently used on constant_exprt typed arguments." somewhat surprising, but if that's the case then of course this makes sense.

**Comment by kroening** [comment]
@tautschnig There's the somewhat stylistic point here whether you want the caller to check the interface at runtime or let the compiler enforce it.

I am in favour of compile-time type checking (I even like predicate subtyping).

A more extreme version of this PR would remove the `exprt` variant, as calling with anything but `constant_exprt` will never work.

**Comment by tautschnig** [comment]
> There's the somewhat stylistic point here whether you want the caller to check the interface at runtime or let the compiler enforce it.

Yes, but that was not something that the commit message talked about.

> I am in favour of compile-time type checking (I even like predicate subtyping).

Certainly.

> A more extreme version of this PR would remove the `exprt` variant, as calling with anything but `constant_exprt` will never work.

The current approach, however, avoids a lot of redundancy: instead of first checking in each an every place, just call the function. In places where we do know that it is a constant _and_ do expect that we have a numeric constant we ought to be using `numeric_cast_v`.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 3437e52b4b57abefce9109ec282a694b7829f201).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99243396
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**